### DISCUSSION TITLE
[@types/wicg-file-system-access] Methods defined in the global scope instead of being defined on Window

### DISCUSSION THREAD

**Discussion by guillaumebrunerie** [DefinitelyTyped/DefinitelyTyped#68451] [Issues with a @types package]
This package defines new types/methods in the global scope (with `declare global { ... }`). I believe this is incorrect, and they should instead be defined in the `Window` interface.
For instance `showDirectoryPicker` (https://developer.mozilla.org/en-US/docs/Web/API/Window/showDirectoryPicker) is supposed to be in the `Window` interface, not in the global scope.

[a comment by an automated account is not shown]

**Comment by RReverser** [comment]
Well, in the browser those are the same. It doesn't make much sense to require them to be only available via e.g. `window.showOpenFilePicker()` when it's valid and more common to call them as just `showOpenFilePicker()`.

**Reply by guillaumebrunerie** [reply]
My use case is that I have an app with several windows and I want to make sure that I call `showDirectoryPicker` on the right window object (otherwise it doesn’t work as the user is not interacting with the right window). But `window.open` returns a value of type `Window` which does not have the `showDirectoryPicker` method, so I need to cast it to `Window & typeof globalThis` to make it typecheck (or get the proper window object from somewhere else).

> It doesn't make much sense to require them to be only available via e.g. `window.showOpenFilePicker()` when it's valid and more common to call them as just `showOpenFilePicker()`.

Hmm, I thought that having them on the window object would make them globally available as well, but apparently not. I looked at several other APIs and it seems a bit inconsistent:
- `getScreenDetails` from `@types/webscreens-window-placement` is only available on `Window`, not globally.
- `ResizeObserver` is only available globally, not on `Window` (despite not working properly if called on the wrong window).
- `requestIdleCallback` is available both globally and on `Window`, but it looks like they needed to duplicate the type declaration.

**Reply by guillaumebrunerie** [reply]
I guess it could be fixed without too much duplication by simply adding
```ts
interface Window {
    showOpenFilePicker: typeof showOpenFilePicker;
    showSaveFilePicker: typeof showSaveFilePicker;
    showDirectoryPicker: typeof showDirectoryPicker;
}
```

**Reply by RReverser** [reply]
> My use case is that I have an app with several windows and I want to make sure that I call showDirectoryPicker on the right window object

Ohh interesting use-case. Does that even work? I didn't think it's possible to call those APIs on other windows, as they require user interaction, so I assumed they're always limited to current window only.

**Reply by guillaumebrunerie** [reply]
>Ohh interesting use-case. Does that even work? I didn't think it's possible to call those APIs on other windows, as they require user interaction, so I assumed they're always limited to current window only.

Yes, it works fine, as long as you call those APIs on the window that the user interacted with. That's one of the tricky bits when working with several windows, you have to understand the difference between `window` (which is the window containing the code in question, I think) and the window from which the code is run (which may be a different one, for instance if it is in an event listener set on a DOM element in a different window).
I couldn’t make a codepen/codesandbox, as showing file pickers is blocked in cross-origin iframes, apparently, but here is below an example that you can run directly on your computer by creating `index.html` and `index.js` and opening `index.html` in Chrome. You'll see that from the popup, only `popup.showDirectoryPicker` works, not `showDirectoryPicker`. And in order to get event listeners that work in any window, you need to do some weird gymnastics like
```js
const currentWindow = event.target.ownerDocument.defaultView;
currentWindow.showDirectoryPicker();
```

This is not limited to file system APIs, there is a similar issue with `ResizeObserver` for instance, which does not work correctly in popup windows if called simply as `new ResizeObserver(...)` (see https://codepen.io/Latcarf/pen/RwdyLXr if you are curious).

```html
<!-- index.html -->

<html>
  <body>
    <script src="./index.js"></script>
  </body>
</html>
```

```js
// index.js

let popup;

const addButton = (document, label, callback) => {
  const div = document.createElement("div");
  div.style.marginTop = "5px";
  const button = document.createElement("button");
  button.textContent = label;
  button.addEventListener("click", callback);
  document.body.appendChild(div);
  div.appendChild(button);
};

addButton(document, "Open popup", () => {
  popup = window.open("", "", "left=200,top=200,width=400,height=200");
  addButton(popup.document, "showDirectoryPicker", () => showDirectoryPicker());
  addButton(popup.document, "window.showDirectoryPicker", () => window.showDirectoryPicker());
  addButton(popup.document, "popup.showDirectoryPicker", () => popup.showDirectoryPicker());
  addButton(popup.document, "(current window).showDirectoryPicker", (event) => {
    const currentWindow = event.target.ownerDocument.defaultView;
    currentWindow.showDirectoryPicker();
  });
});

addButton(document, "showDirectoryPicker", () => showDirectoryPicker());
addButton(document, "window.showDirectoryPicker", () => window.showDirectoryPicker());
addButton(document, "popup.showDirectoryPicker", () => popup.showDirectoryPicker());
addButton(document, "(current window).showDirectoryPicker", (event) => {
  const currentWindow = event.target.ownerDocument.defaultView;
  currentWindow.showDirectoryPicker();
});
```
